keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -0,0 +1,626 @@
+
+#### Modifying Consumers of EndpointAddress API to Use Dual-Stack Addresses
+Any clients that consume the [Kubernetes EndpointAddress API](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#endpointaddress-v1-core) will need to be modified to read the new plural IPs field, if they are to take advantage of the dual-stack feature. This proposal will include dual-stack updates to the following consumers of Kubernetes endpoints:
+- kube-proxy, iptables mode
+- kube-proxy, IPVS mode
+- Nginx Ingress controller
+Other users/consumers of this API may be updated to dual-stack over time, but that work is considered outside of the scope of this proposal.
+
+#### 'kubectl get endpoints' Command Display for Dual-Stack Backend Pods
+The 'kubectl get endpoints ...' command display should be changed so that:
+- IPv4/IPv6 endpoint address pairs appear within curly braces.
+- If dual-stack endpoint address pairs are present, then each pair should be printed on its own line as show below.
+Example:
+```
+       kube-master# kubectl get endpoints
+       NAME            ENDPOINTS                             AGE
+       kubernetes      {[fd00::100]:6443,10.0.0.2:6643}      15m
+       nginx-service   {[fd00:db8:1::2]:80,192.168.1.3:80}   20m
+                       {[fd00:db8:2::2]:80,192.168.2.3:80}   20m
+       kube-master#
+```
+
+#### 'kubectl describe service' Command Display for Dual-Stack Backend Pods
+The 'kubectl describe service ...' command display should be changed so that:
+- IPv4/IPv6 endpoint address pairs appear within curly braces.
+- If dual-stack endpoint address pairs are present, then each pair should be printed on its own line as show below.
+Example:
+```
+       kube-master# kubectl describe service nginx-service
+       .
+       .
+       .
+       Endpoints:   {[fd00:db8:1::2]:80,192.168.1.3:80}
+                    {[fd00:db8:2::2]:80,192.168.2.3:80}
+       .
+       .
+       .
+       kube-master#
+```
+
+### kube-proxy Operation
+
+Kube-proxy will be modified to drive iptables and ip6tables in parallel. This will require the implementation of a second "proxier" interface in the Kube-Proxy server in order to modify and track changes to both tables. This is required in order to allow exposing services via both IPv4 and IPv6, e.g. using Kubernetes:
+  - NodePort
+  - ExternalIPs
+
+#### Kube-Proxy Startup Configuration Changes
+
+##### Multiple bind addresses configuration
+The existing "--bind-address" option for the [kube-proxy startup configuration](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-proxy/) will be modified to support multiple IP addresses in a comma-separated list (rather than a single IP string).
+```
+  --bind-address  stringSlice   (IP addresses, in a comma separated list, Default: [0.0.0.0,])
+```
+Only the first address of each IP family will be used; all others will be ignored.
```

Comment:
We should probably change this to CIDRs and assume that "no size portion" == "/32"

Reply:
That could be a PR of its own